File: leetcode/257-binary-tree-paths/main.go
package main

import "strconv"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
	2nd approach: iterative dfs

	Time  O(n)
	Space O(h)
	24 ms, faster than 76.92%
*/

type StackItem struct {
	Node *TreeNode
	Path string
}

func binaryTreePaths(root *TreeNode) []string {
	if root == nil {
		return []string{}
	}
	stack := []*StackItem{}
	stack = append(stack, &StackItem{root, ""})
	res := []string{}
	for len(stack) > 0 {
		pop := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		node, path := pop.Node, pop.Path
		newPath := path + "->" + strconv.Itoa(node.Val)
		if node.Left == nil && node.Right == nil {
			res = append(res, newPath[2:])
		}
		if node.Left != nil {
			stack = append(stack, &StackItem{node.Left, newPath})
		}
		if node.Right != nil {
			stack = append(stack, &StackItem{node.Right, newPath})
		}
	}
	return res
}

func main() {

}

File: leetcode/257-binary-tree-paths/main.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def __init__(self):
        self.result = []

    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]

        1st approach: recursive dfs

        Time  O(n)
        Space O(h)
        24 ms, faster than 76.92%
        """
        self.dfs(root, "")
        return self.result

    def dfs(self, node, path):
        if node == None:
            return
        newPath = path+"->"+str(node.val)
        if node.left == None and node.right == None:
            self.result.append(newPath[2:])
        self.dfs(node.left, newPath)
        self.dfs(node.right, newPath)
